resources/type-system/upper-lower-bounds.md

```diff
@@ -0,0 +1,265 @@
+
+## Upper bounds
+
+We define the upper bound of two types T1 and T2 to be **UP**(T1,T2) as follows.
+
+
+- **UP**(`T`, `T`) = `T`
+- **UP**(`T1`, `T2`) = `T1` if:
+  - **TOP**(`T1`) and **TOP**(`T2`)
+  - and  **MORETOP**(`T1`, `T2`)
+- **UP**(`T1`, `T2`) = `T2` if:
+  - **TOP**(`T1`) and **TOP**(`T2`)
+  - and  not **MORETOP**(`T1`, `T2`)
+- **UP**(`T1`, `T2`) = `T1` if **TOP**(`T1`)
+- **UP**(`T1`, `T2`) = `T1` if **BOTTOM**(`T2`)
+- **UP**(`T1`, `T2`) = `T2` if **TOP**(`T2`)
+- **UP**(`T1`, `T2`) = `T2` if **BOTTOM**(`T1`)
+
+- **UP**(`X1 extends B1`, `T2`) = 
+  - `T2` if `X1 <: T2`
+  - otherwise `X1` if `T2 <: X1`
+  - otherwise **UP**(`B1[Object/X1]`, `T2`)
+
+- **UP**(`X1 & B1`, `T2`) = 
+  - `T2` if `X1 <: T2`
+  - otherwise `X1` if `T2 <: X1`
+  - otherwise **UP**(`B1[Object/X1]`, `T2`)
+
+- **UP**(`T1`, `X2 extends B2`) = 
+  - `X2` if `T1 <: X2`
+  - otherwise `T1` if `X2 <: T1`
+  - otherwise **UP**(`T1`, `B2[Object/X2]`)
+
+- **UP**(`T1`, `X2 & B2`) = 
+  - `X2` if `T1 <: X2`
+  - otherwise `T1` if `X2 <: T1`
+  - otherwise **UP**(`T1`, `B2[Object/X2]`)
+
+- **UP**(`T Function<...>(...)`, `Function`) = `Function`
+- **UP**(`T Function<...>(...)`, `T2`) = `Object`
+- **UP**(`Function`, `T Function<...>(...)`) = `Function`
+- **UP**(`T1`, `T Function<...>(...)`) = `Object`
+
+- **UP**(`T0 Function<X0 extends B00, ... Xm extends B0m>(P00, ... P0k)`,
+         `T0 Function<X0 extends B10, ... Xm extends B1m>(P10, ... P1l)` = 
+   `R0 Function<X0 extends B20, ..., Xm extends B2m>(P20, ..., P2q)` if:
+     - each `B0i` and `B1i` are equal types (syntactically)
+     - Both have the same number of required positional parameters
+     - `q` is min(`k`, `l`)
+     - `R0` is **UP**(`T0`, `T1`)
+     - `B2i` is `B0i`
+     - `P2i` is **DOWN**(`P0i`, `P1i`)
+- **UP**(`T0 Function<X0 extends B00, ... Xm extends B0m>(P00, ... P0k, Named0)`,
+         `T0 Function<X0 extends B10, ... Xm extends B1m>(P10, ... P1k, Named1)` = 
+   `R0 Function<X0 extends B20, ..., Xm extends B2m>(P20, ..., P2k, Named2)` if:
+     - each `B0i` and `B1i` are equal types (syntactically)
+     - All positional parameters are required
+     - `R0` is **UP**(`T0`, `T1`)
+     - `B2i` is `B0i`
+     - `P2i` is **DOWN**(`P0i`, `P1i`)
+     - `Named2` contains exactly `R2i xi` for each `xi` in both `Named0` and `Named1` 
+        - where `R0i xi` is in `Named0`
+        - where `R1i xi` is in `Named1`
+        - and `R2i` is **DOWN**(`R0i`, `R1i`)
+
+- **UP**(`T Function<...>(...)`, `S Function<...>(...)`) = `Function` otherwise
+- **UP**(`T1`, `T2`) = `T2` if `T1` <: `T2`
+  - Note that both types must be interface types at this point
+- **UP**(`T1`, `T2`) = `T1` if `T2` <: `T1`
+  - Note that both types must be interface types at this point
+- **UP**(`C<T0, ..., Tn>`, `C<S0, ..., Sn>`) = `C<R0,..., Rn>` where `Ri` is **UP**(`Ti`, `Si`)
+- **UP**(`C0<T0, ..., Tn>`, `C1<S0, ..., Sk>`) = least upper bound of two interfaces
+  as in Dart 1.
```

Comment:
No.

Reply:
The rules for this case (current DartLangSpecDraft.pdf Sect. 19.10.2) refer to 'the set of superinterfaces' of an interface. This must mean the transitive closure of the 'direct superinterface' relation, and it must apply to parameterized types (e.g., not just to class declarations).

But I can't see anything in the spec to support a claim that it includes all supertypes. Also, that doesn't have to be a finite set, and it isn't a very implementation-friendly interpretation. For instance, with `class A1 {}`, `class A2 {}`, and `class B implements A1, A2 {}`, `Iterable<void Function(B)>` is in the intersection of the supertypes of `Iterable<void Function(A1)>` and `Iterable<void Function(A2)>`, but we wouldn't require an implementation to find it.

So, with `class A {}` and `class B1 implements A {}`, the type `Iterable<B1>` has superinterfaces `Iterable<B1>` and `Object`, but not, say, `Iterable<A>`.

However, the current implementations (`dart` and `dartanalyzer`) have no issues with the following:

```dart
class A {}
class B1 implements A {}
class B2 implements A {}

main() {
  bool b = true;
  Iterable<B1> it1;
  Iterable<B2> it2;
  var it = b ? it1 : it2;
  Expect.assertType<Iterable<A>>(GetType(it).type);
}
```

This implies that the algorithm is _actually_ applied recursively on the type arguments, rather than just taking the intersection of the sets of superinterfaces as specified.

We can see that it runs the algorithm (involving non-class types), too:

```dart
class A {}
class B implements A {}

main() {
  bool b = true;
  Iterable<void Function(B)> it1;
  Iterable<void Function(A)> it2;
  var it = b ? it1 : it2;
  Expect.assertType<Iterable<void Function(B)>>(GetType(it).type);
}
```

So I really find it difficult to describe the current approach if we insist that it isn't 'invoking rules stated here as needed in recursive steps'.